File: tools/SharedJavaLibrary/src/network/message/RemoteMessage.java
package network.message;

public class RemoteMessage {

  public RemoteMessage(int type, byte [] data) {
    this.type = type;
    this.data = data;
  }
  
  public byte [] getData() {
    return data;
  }
  
  protected int getTypeID() {
    return type;
  }
  
  protected byte [] data;
  
  private int type;
  
  public static int readInt(byte [] data, int offset) {
    if (offset + 3 >= data.length) {
      return 0;
    }
    
//    System.out.printf("%x %x %x %x\n", (int)data[offset], (int)data[offset + 1], (int)data[offset + 2], (int)data[offset + 3]);
    return ((int)data[offset] & 0xFF) << 24 |
           ((int)data[offset + 1] & 0xFF) << 16 |
           ((int)data[offset + 2] & 0xFF) << 8  |
           ((int)data[offset + 3] & 0xFF);
  }

  public static float readFloat(byte [] data, int offset) {
    if (offset + 3 >= data.length) {
      return 0;
    }
    
    int intBits = readInt(data, offset);
    
//    System.out.println(Integer.toHexString(intBits));
    return Float.intBitsToFloat(intBits);
  }
  
  public static boolean writeInt(byte [] data, int offset, int value) {
    if (offset + 3 >= data.length) {
      return false;
    }
    
    data[offset]     = (byte)((value >> 24) & 0xFF);
    data[offset + 1] = (byte)((value >> 16) & 0xFF);
    data[offset + 2] = (byte)((value >> 8) & 0xFF);
    data[offset + 3] = (byte)(value & 0xFF);
    
    return true;
  }
  
  public static boolean writeFloat(byte [] data, int offset, float value) {
    int intBits = Float.floatToIntBits(value);
    
    return writeInt(data, offset, intBits);
  }

}

File: robot/Comm/Messages/Remote/RemoteMessage.cpp
#include <cstring>
#include <stdint.h>
#include <netinet/in.h>

#include "RemoteMessage.h"

RemoteMessage::RemoteMessage(int _type, unsigned int _size, void const * _data)
  : size(0),
    data(NULL),
    type(_type) {
  if ((_size > 0) && (_data != NULL)) {
    size = _size;
    data = new char[_size];
    memcpy(data, _data, _size);
  }
}

RemoteMessage::~RemoteMessage() {
  if (data != NULL) {
    delete [] data;
  }
}

unsigned int RemoteMessage::getSize() const {
  return size;
}

void const * RemoteMessage::getData() const {
  return data;
}

int RemoteMessage::getTypeID() const {
  return type;
}

void RemoteMessage::writeInt(char *data, int value) {
  uint32_t valueNetwork = htonl(value);
  memcpy(data, &valueNetwork, 4);
}

void RemoteMessage::writeFloat(char *data, float value) {
  uint32_t *floatBitsPointer = reinterpret_cast<uint32_t *>(&value);
  uint32_t floatBits = htonl(*floatBitsPointer);
  memcpy(data, &floatBits, 4);
}

int RemoteMessage::readInt(char const *data) {
  uint32_t valueNetwork(0);
  memcpy(&valueNetwork, data, 4);

  return ntohl(valueNetwork);
}

float RemoteMessage::readFloat(char const *data) {
  uint32_t valueNetwork(0);
  memcpy(&valueNetwork, data, 4);
  valueNetwork = ntohl(valueNetwork);
  float *floatBitsPointer = reinterpret_cast<float *>(&valueNetwork);

  return *floatBitsPointer;
}

